schema/migration.go

```diff
@@ -37,57 +47,425 @@ func (x Versions) Latest() Version { return x[len(x)-1] }
+		return b
+	case UUIDType, DateTimeType, ArrayType, ObjectType:
+		// non convertible
+	}
+	return nil
+}
+
+func convertFromString(toType FieldType, val string, maxLength int) any {
+	switch toType {
+	case BoolType:
+		if val == "true" {
+			return true
+		}
+		if val == "false" {
+			return false
+		}
+	case Int32Type:
+		if i, err := strconv.ParseInt(val, 10, 32); err == nil {
+			return i
+		}
+	case Int64Type:
+		if i, err := strconv.ParseInt(val, 10, 64); err == nil {
+			return i
+		}
+	case DoubleType:
+		if f, err := strconv.ParseFloat(val, 64); err == nil {
+			return f
+		}
+	case StringType:
+		if maxLength > 0 && len(val) > maxLength {
+			return val[:maxLength]
+		}
+		return val
+	case ByteType:
+		if maxLength > 0 && len(val) > maxLength {
+			return []byte(val[:maxLength])
+		}
+		return []byte(val)
+	case UUIDType:
+		if _, err := uuid.Parse(val); err == nil {
+			return val
+		}
+	case DateTimeType:
+		// TODO: Detect more date time formats
+		if _, err := time.Parse(time.RFC3339Nano, val); err == nil {
+			return val
+		}
+	case ArrayType, ObjectType:
+		// non convertible
+	}
+	return nil
+}
+
+func convertType(change *VersionDeltaField, val any) any {
+	switch v := val.(type) {
+	case bool:
+		return convertFromBool(change.To, v)
+	case string:
+		if change.From == ByteType {
+			if b, err := base64.StdEncoding.DecodeString(v); err == nil {
+				return convertFromBytes(change.To, b, change.MaxLength)
+			}
+		} else {
+			return convertFromString(change.To, v, change.MaxLength)
+		}
+	case json.Number: // comes from initial document unmarshal
+		switch change.From {
+		case Int64Type, Int32Type:
+			if i, err := v.Int64(); err == nil {
+				return convertFromInt64(change.To, i)
+			}
+		case DoubleType:
+			if i, err := v.Float64(); err == nil {
+				return convertFromDouble(change.To, i)
+			}
+		}
+	case float64: // float64, int64 comes from our conversion
+		return convertFromDouble(change.To, v)
+	case int64:
+		return convertFromInt64(change.To, v)
+	}
+
+	return nil
+}
+
+// applyArrayFieldDelta applies delta to every array element,
+// recursively descending multidimensional arrays.
+// If array element is not convertible to target type function returns false,
+// so the caller can remove the array entirely.
+func applyArrayFieldDelta(arr []any, change *VersionDeltaField, i int) bool {
+	for key, v := range arr {
+		switch m := v.(type) {
+		case map[string]any:
+			if !applyObjectFieldDelta(m, change, i+1) {
+				delete(m, change.KeyPath[i])
+				return false
+			}
+		case []any:
+			if !applyArrayFieldDelta(m, change, i+1) {
+				return false
+			}
+		default:
+			arr[key] = convertType(change, m)
+			if arr[key] == nil {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func applyPrimitiveFieldDelta(doc map[string]any, change *VersionDeltaField) bool {
+	key := change.KeyPath[len(change.KeyPath)-1]
+
+	// array is not convertible to anything
+	if _, ok := doc[key].([]any); ok {
+		delete(doc, key)
```

Comment:
At this point are we sure that this key is from the old type? Or do we need to check the type of array?